docs/createrepo.8
.TH "createrepo" "8" "2005 Jan 2" "Seth Vidal" ""

.SH "NAME"
createrepo \- Create repomd (xml-rpm-metadata) repository

.SH "SYNOPSIS"
\fBcreaterepo\fP [options] <directory>
.PP

.SH "DESCRIPTION"
\fBcreaterepo\fP is a program that creates a repomd (xml-based rpm metadata) repository from a set of rpms.

.SH "OPTIONS"
.IP "\fB\-u --baseurl\fP <url>"
Optional base URL location for all files.
.IP "\fB\-o --outputdir\fP <url>"
Optional output directory (useful for read only media).
.IP "\fB\-x --excludes\fP <package>"
File globs to exclude, can be specified multiple times.
.IP "\fB\-i --pkglist\fP <filename>"
specify a text file which contains the complete list of files to
include in the repository from the set found in the directory. File format is one
package per line, no wildcards or globs.
.IP "\fB\-n --includepkg\fP"
specify pkgs to include on the command line. Takes urls as well as local paths.
.IP "\fB\-q --quiet\fP"
Run quietly.
.IP "\fB\-g --groupfile\fP <groupfile>"
A precreated xml filename to point to for group information.
.br
See examples section below for further explanation.
.IP "\fB\-v --verbose\fP"
Run verbosely.
.IP "\fB\-c --cachedir\fP <path>"
Specify a directory to use as a cachedir. This allows createrepo to create a
cache of checksums of packages in the repository. In consecutive runs of
createrepo over the same repository of files that do not have a complete
change out of all packages this decreases the processing time dramatically.
.br
.IP "\fB\--basedir\fP"
Basedir for path to directories in the repodata, default is the current working
directory.
.br
.IP "\fB\--update\fP"
If metadata already exists in the outputdir and an rpm is unchanged
(based on file size and mtime) since the metadata was generated, reuse
the existing metadata rather than recalculating it. In the case of a
large repository with only a few new or modified rpms this can
significantly reduce I/O and processing time.
.br
.IP "\fB\--skip-stat\fP"
skip the stat() call on a --update, assumes if the filename is the same
then the file is still the same (only use this if you're fairly trusting or
gullible).
.br
.IP "\fB\--update-md-path\fP"
Use the existing repodata for --update, from this path.
.br
.IP "\fB\-C --checkts\fP"
Don't generate repo metadata, if their timestamps are newer than its rpms.
This option decreases the processing time drastically again, if you happen
to run it on an unmodified repo, but it is (currently) mutual exclusive
with the --split option. NOTE: This command will not notice when 
packages have been removed from repo. Use --update to handle that.
.br
.IP "\fB\--split\fP"
Run in split media mode. Rather than pass a single directory, take a set of
directories corresponding to different volumes in a media set.
.br
.IP "\fB\-p --pretty\fP"
Output xml files in pretty format.
.IP "\fB\--version\fP"
Output version.
.IP "\fB\-h --help\fP"
Show help menu.

.IP "\fB\-d --database\fP"
Generate sqlite databases for use with yum. This is now the default.

.IP "\fB\--no-database\fP"
Do not generate sqlite databases in the repository.

.IP "\fB\-S --skip-symlinks\fP"
Ignore symlinks of packages
.IP "\fB\-s --checksum\fP"
Choose the checksum type used in repomd.xml and for packages in the metadata.
The default is now "sha256" (if python has hashlib). The older default was
"sha", which is actually "sha1", however explicitly using "sha1" doesn't work
on older (3.0.x) versions of yum, you need to specify "sha".
.IP "\fB\--profile\fP"
Output time based profiling information.
.IP "\fB\--changelog-limit\fP CHANGELOG_LIMIT"
Only import the last N changelog entries, from each rpm, into the metadata
.IP "\fB\--unique-md-filenames\fP"
Include the file's checksum in the metadata filename, helps HTTP caching (default)

.IP "\fB\--simple-md-filenames\fP"
Do not include the file's checksum in the metadata filename.

.IP "\fB\--retain-old-md\fP"
Keep around the latest (by timestamp) N copies of the old repodata (so clients
with older repomd.xml files can still access it). Default is 0.

.IP "\fB\--distro\fP"
Specify distro tags. Can be specified more than once. Optional syntax specifying a
cpeid(http://cpe.mitre.org/) --distro=cpeid,distrotag
.IP "\fB\--content\fP"
Specify keyword/tags about the content of the repository. Can be specified more than once.
.IP "\fB\--repo\fP"
Specify keyword/tags about the repository itself. Can be specified more than once.
.IP "\fB\--revision\fP"
Arbitrary string for a repository revision.
.IP "\fB\--deltas\fP"
Tells createrepo to generate deltarpms and the delta metadata
.IP "\fB\--oldpackagedirs\fP PATH"
paths to look for older pkgs to delta against. Can be specified multiple times
.IP "\fB\--num-deltas\fP int"
the number of older versions to make deltas against. Defaults to 1
.IP "\fB\--read-pkgs-list\fP READ_PKGS_LIST
output the paths to the pkgs actually read useful with  --update
.IP "\fB\--max-delta-rpm-size\fP MAX_DELTA_RPM_SIZE
max size of an rpm that to run deltarpm against (in bytes)
.IP "\fB\--workers\fP WORKERS
number of workers to spawn to read rpms
.IP "\fB\--compress-type\fP
specify which compression method to use: compat (default),
xz (may not be available), gz, bz2.
.IP

.SH "EXAMPLES"
Here is an example of a repository with a groups file. Note that the
groups file should be in the same directory as the rpm packages
(i.e. /path/to/rpms/comps.xml).
.br
.PP
\fBcreaterepo\fP \-g comps.xml /path/to/rpms

.SH "FILES"
.nf
repodata/filelists.xml.gz
repodata/other.xml.gz
repodata/primary.xml.gz
repodata/repomd.xml 
.fi
.PP 
.SH "SEE ALSO"
.I yum (8) yum.conf (5)

.PP 
.SH "AUTHORS"
.nf 
See the Authors file
.fi 

.PP 
.SH "BUGS"
Any bugs which are found should be emailed to the mailing list:
rpm-metadata@lists.baseurl.org
or reported in trac at: http://createrepo.baseurl.org
.fi


ChangeLog
2011-01-26  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo.spec, createrepo/__init__.py: mark as 0.9.9

2011-01-26  Seth Vidal <skvidal@fedoraproject.org>

	Merge branch 'master' of
	ssh://createrepo.baseurl.org/srv/projects/createrepo/git/createrepo
	* 'master' of
	ssh://createrepo.baseurl.org/srv/projects/createrepo/git/createrepo:
	Override timestamp check on repos. for mergerepo (like repodiff).
	Add createrepo --workers (non)completion.   Add modifyrepo option
	completion.

2011-01-21  James Antill <james@and.org>

	* createrepo/merge.py: Override timestamp check on repos. for
	mergerepo (like repodiff).

2011-01-03  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py: make sure when we want to look for rpms we
	say .rpm not rpm b/c with the latter we catch .drpm files, too. :(

2010-11-02  Ville Skyttä <ville.skytta@iki.fi>

	* createrepo.bash: Add createrepo --workers (non)completion.

2010-11-02  Ville Skyttä <ville.skytta@iki.fi>

	* createrepo.bash: Add modifyrepo option completion.

2010-10-08  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo.spec, createrepo/__init__.py: - add yum 3.2.29 requirement b/c of the small change I needed to
	repoMDObject.py -  set it to use /usr/share/createrepo/worker.py

2010-10-07  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py: remove libxml2 import from __init__.py :)


2010-10-07  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py: make createrepo use the repomd/repodata
	mechanism from yum for making a repomd.xml which simplifies the code
	dramatically since we don't have to mess with xml in here.

2010-10-07  Seth Vidal <skvidal@fedoraproject.org>

	* modifyrepo.py: fix up the usage output for modifyrepo

2010-09-10  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py, worker.py: - make sure we handle remote_url pkgs correctly until we get the
	worker hooked up to handle them - if there are no pkgs to handle,
	don't launch workers with nothing to do. - give better output from
	the workers and have them obey -v/-q - everyone loves callbacks!

2010-09-09  Seth Vidal <skvidal@fedoraproject.org>

	* Makefile, createrepo/__init__.py, createrepo/utils.py,
	createrepo/yumbased.py, genpkgmetadata.py, worker.py: create a
	worker script for createrepo so createrepo can fork off N processes
	to handle the md gathering from pkgs. This should speed up results
	on systems which have been cpubound on the createrepo process.  If
	you're io bound it won't help you at all, and MAY make it worse.
	many misc issues to iron out here - not the least of which is the
	callback output and gathering stdout/stderr from the workers

2010-08-20  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py: handle broken locking on nfs target dirs
	better if database is true.  - sqlite dbs don't like being made on
	locations without locking available. - if we know we're going to be
	creating dbs then we should attempt to lock before doing   anything
	else and bail out nicely if we can't get an exclusive lock

2010-08-19  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo.spec: require yum 3.2.28 due to the imports in
	modifyrepo

2010-08-19  Seth Vidal <skvidal@fedoraproject.org>

	* docs/modifyrepo.1: document --mdtype option

2010-08-19  Seth Vidal <skvidal@fedoraproject.org>

	* modifyrepo.py: - add option parsing for --mdtype - use the yum repomd objects to
	read/write the repomd.xml - add mdtype option to RepoMetadata.add()
	method

2010-06-11  Seth Vidal <skvidal@fedoraproject.org>

	Merge branch 'master' of
	ssh://createrepo.baseurl.org/srv/projects/createrepo/git/createrepo
	* 'master' of
	ssh://createrepo.baseurl.org/srv/projects/createrepo/git/createrepo:
	Don't use /usr/bin/env ... it's evil   --database is now the default
	for mergerepo too, have --no-database in completions instead.

2010-06-11  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py: - add option to createrepo config to collapse libc.so.6 requires -
	this will only work with yum 3.2.28 and beyond

2010-06-03  James Antill <james@and.org>

	* modifyrepo.py: Don't use /usr/bin/env ... it's evil

2010-06-02  Ville Skyttä <ville.skytta@iki.fi>

	* createrepo.bash: --database is now the default for mergerepo too,
	have --no-database in completions instead.

2010-06-01  Seth Vidal <skvidal@fedoraproject.org>

	* mergerepo.py: whoops - no-database shouldn't default to true!

2010-06-01  Seth Vidal <skvidal@fedoraproject.org>

	* mergerepo.py: add --no-database to mergrepo, too

2010-05-31  Ville Skyttä <ville.skytta@iki.fi>

	* createrepo.bash: --database is now the default, have --no-database
	in completions instead.

2010-05-28  Seth Vidal <skvidal@fedoraproject.org>

	* docs/createrepo.8: update the docs for --no-database

2010-05-28  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py, genpkgmetadata.py: make -d/--database  the
	default  add --no-database in case someone somewhere needs to do
	that

2010-04-26  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py: fixme comments about try/excepting the
	database creation calls due to a weird issue with locks not working
	on a nfs mount and createreepo tracing back with a TypeError of all
	things

2010-04-21  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py, createrepo/readMetadata.py: if we cannot
	find one of the old repodata files make the warning look more dire
	make sure we look for the 'repodata' subdir inside update_md_path

2010-04-21  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py: when the update_md_path doesn't exist -
	emit a warning of some kind - rather than a somewhat quieter message
	from MetadataIndex()  this is mostly to help jesse b/c he asked
	nicely

2010-04-16  Colin Walters <walters@fedoraproject.org>

	* genpkgmetadata.py: if we're not a tty, don't use the progress
	output

2010-04-15  Seth Vidal <skvidal@fedoraproject.org>

	Merge branch 'master' of

createrepo.bash
# bash completion for createrepo and friends

_cr_compress_type()
{
    COMPREPLY=( $( compgen -W "$( ${1:-createrepo} --compress-type=FOO / 2>&1 \
        | sed -ne 's/,/ /g' -ne 's/.*[Cc]ompression.*://p' )" -- "$2" ) )
}

_cr_checksum_type()
{
    COMPREPLY=( $( compgen -W 'md5 sha1 sha256 sha512' -- "$1" ) )
}

_cr_createrepo()
{
    COMPREPLY=()

    case $3 in
        --version|-h|--help|-u|--baseurl|--distro|--content|--repo|\
        --revision|-x|--excludes|--changelog-limit|--max-delta-rpm-size)
            return 0
            ;;
        --basedir|-c|--cachedir|--update-md-path|-o|--outputdir|\
        --oldpackagedirs)
            COMPREPLY=( $( compgen -d -- "$2" ) )
            return 0
            ;;
        -g|--groupfile)
            COMPREPLY=( $( compgen -f -o plusdirs -X '!*.xml' -- "$2" ) )
            return 0
            ;;
        -s|--checksum)
            _cr_checksum_type "$2"
            return 0
            ;;
        -i|--pkglist|--read-pkgs-list)
            COMPREPLY=( $( compgen -f -o plusdirs -- "$2" ) )
            return 0
            ;;
        -n|--includepkg)
            COMPREPLY=( $( compgen -f -o plusdirs -X '!*.rpm' -- "$2" ) )
            return 0
            ;;
        --retain-old-md)
            COMPREPLY=( $( compgen -W '0 1 2 3 4 5 6 7 8 9' -- "$2" ) )
            return 0
            ;;
        --num-deltas)
            COMPREPLY=( $( compgen -W '1 2 3 4 5 6 7 8 9' -- "$2" ) )
            return 0
            ;;
        --workers)
            local min=2 max=$( getconf _NPROCESSORS_ONLN 2>/dev/null )
            [[ -z $max || $max -lt $min ]] && max=$min
            COMPREPLY=( $( compgen -W "{1..$max}" -- "$2" ) )
            return 0
            ;;
        --compress-type)
            _cr_compress_type "$1" "$2"
            return 0
            ;;
    esac

    if [[ $2 == -* ]] ; then
        COMPREPLY=( $( compgen -W '--version --help --quiet --verbose --profile
            --excludes --basedir --baseurl --groupfile --checksum --pretty
            --cachedir --checkts --no-database --update --update-md-path
            --skip-stat --split --pkglist --includepkg --outputdir
            --skip-symlinks --changelog-limit --unique-md-filenames
            --simple-md-filenames --retain-old-md --distro --content --repo
            --revision --deltas --oldpackagedirs --num-deltas --read-pkgs-list
            --max-delta-rpm-size --workers --compress-type' -- "$2" ) )
    else
        COMPREPLY=( $( compgen -d -- "$2" ) )
    fi
} &&
complete -F _cr_createrepo -o filenames createrepo genpkgmetadata.py

_cr_mergerepo()
{
    COMPREPLY=()

    case $3 in
        --version|-h|--help|-a|--archlist)
            return 0
            ;;
        -r|--repo|-o|--outputdir)
            COMPREPLY=( $( compgen -d -- "$2" ) )
            return 0
            ;;
        --compress-type)
            _cr_compress_type "" "$2"
            return 0
            ;;
    esac

    COMPREPLY=( $( compgen -W '--version --help --repo --archlist --no-database
        --outputdir --nogroups --noupdateinfo --compress-type' -- "$2" ) )
} &&
complete -F _cr_mergerepo -o filenames mergerepo mergerepo.py

_cr_modifyrepo()
{
    COMPREPLY=()

    case $3 in
        --version|-h|--help|--mdtype)
            return 0
            ;;
        --compress-type)
            _cr_compress_type "" "$2"
            return 0
            ;;
        -s|--checksum)
            _cr_checksum_type "$2"
            return 0
            ;;
    esac

    if [[ $2 == -* ]] ; then
        COMPREPLY=( $( compgen -W '--version --help --mdtype --remove
            --compress --compress-type --checksum --unique-md-filenames
            --simple-md-filenames' -- "$2" ) )
        return 0
    fi

    local i argnum=1
    for (( i=1; i < ${#COMP_WORDS[@]}-1; i++ )) ; do
        if [[ ${COMP_WORDS[i]} != -* &&
              ${COMP_WORDS[i-1]} != @(=|--@(md|compress-)type) ]]; then
            argnum=$(( argnum+1 ))
        fi
    done

    case $argnum in
        1)
            COMPREPLY=( $( compgen -f -o plusdirs -- "$2" ) )
            return 0
            ;;
        2)
            COMPREPLY=( $( compgen -d -- "$2" ) )
            return 0
            ;;
    esac
} &&
complete -F _cr_modifyrepo -o filenames modifyrepo modifyrepo.py

# Local variables:
# mode: shell-script
# sh-basic-offset: 4
# sh-indent-comment: t
# indent-tabs-mode: nil
# End:
# ex: ts=4 sw=4 et filetype=sh
